Bug 1336556 - Build ui with neutrino/webpack instead of grunt

caseywilliams opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/2159)
<!-- Reviewable:end -->

wlach commented on docs/ui/installation.rst:
```diff
@@ -1,6 +1,7 @@
 ============
 
+It's possible to work on the UI without setting up the Vagrant VM.
 You can work on the UI without needing a VM, by using web-server.js.
```
Did you mean to delete this line?

eliperelman commented on package.json:
```diff
@@ -9,41 +9,61 @@
   },
   "scripts": {
-    "test": "karma start tests/ui/config/karma.conf.js --single-run --browsers Firefox"
+    "start:stage": "NODE_ENV=development SERVICE_DOMAIN=staging neutrino start",
```
When using Neutrino commands `start`, `build`, and `test`, the `NODE_ENV` is overridden.

- `start` sets `process.env.NODE_ENV = 'development'`
- `build` sets `process.env.NODE_ENV = 'production'`
- `start` sets `process.env.NODE_ENV = 'test'`

These aren't currently overridable.

eliperelman commented on neutrino-preset.js:
```diff
@@ -0,0 +1,207 @@
+'use strict';
```
I think this configuration is pretty good for the current Neutrino. Neutrino v4 should clean this up quite a bit.

My one recommendation would probably be to break this up into a few files, in a directory named `neutrino-preset`. Then your preset would just be the same, `neutrino-preset`.

edmorley commented on ui/js/react/treeherder/revision-list.jsx:
```diff
@@ -0,0 +1,95 @@
+'use strict';
```
Do we need to nest the react files in `ui/js/react/treeherder/revision-list.jsx` for build reasons?
If not, I'm wondering if we want to structure by function rather than react vs not? (Though I'll defer to Cameron on this)

caseywilliams replied:
No reason to separate the react files based on the build process - this can be organized any way at all!

camd replied:
Ed has a good point.  When we had a tiny amount of react, it perhaps made sense to section it off.  But now that we're going to be much more hybrid, putting them in their functional areas probably makes more sense.  Though doesn't need to happen in this PR, IMO.  :)

edmorley replied:
I'd agree about doing in another PR except that this PR is the one that moves them to this new location, so I think we should either (a) leave them where they are and restructure later, (b) still move them, but use a more functional directory structure rather than react vs not. Happy with either approach :-)

caseywilliams commented on docs/common_tasks.rst:
```diff
@@ -123,7 +123,7 @@ To do this:
 * Tell people to visit: ``https://<your-username>.github.io/treeherder/ui/``
 
-There is no need to perform a ``grunt build`` prior. After switching away from the local gh-pages branch, you will need to recreate ``ui/js/config/local.conf.js`` if desired, due to the ``git add -f``.
+There is no need to run ``npm run build`` prior. After switching away from the local gh-pages branch, you will need to recreate ``ui/js/config/local.conf.js`` if desired, due to the ``git add -f``.
```
That sounds right to me - I'll update this

caseywilliams commented on docs/ui/installation.rst:
```diff
@@ -1,48 +1,95 @@
+
+This will run the unminified UI using `<url>` as the service domain.
+
+Running the unminified UI with Vagrant
```
I'll update the main installation doc with some info about how to do this

caseywilliams commented on neutrino-custom/base.js:
```diff
@@ -0,0 +1,176 @@
+        .add(path.join(UI, 'entry-userguide.js'))
+        .end();
+
+    // Include commonly used files in a vendor chunk:
```
That's the kind of behavior that's going on here, but I initially configured it to just bundle anything used twice or more into the vendor chunk, since that seemed to produce the overall smallest bundle. Agreed it would be better to base the division on whether the code comes from vendor libs, as you describe - that's easy to set up - I'll do it today.

caseywilliams commented on package.json:
```diff
@@ -10,40 +10,56 @@
+    "react-dom": "15.3.1",
+    "taskcluster-client": "1.6.3"
   },
   "devDependencies": {
```
I think I've limited this to the libraries that are actually `require`d in the TH app (react doesn't depend on react-dom, nor does ngreact depend on react, for example, but all are necessary to render TH's react components). Or am I misunderstanding?

edmorley replied:
Ah right I didn't look closely enough, I mistakenly thought some were sub-dependencies.

edmorley replied:
Oh actually reading back again now, I think you thought I was talking about `dependencies` whereas I was meaning `devDependencies`? eg: `copy-webpack-plugin` and others will be pulled in from the neutrino-* packages, so don't need to be defined explicitly?
https://github.com/mozilla-neutrino/neutrino-dev/blob/master/packages/neutrino-preset-web/package.json#L20

Also, I've just realised - on Heroku the npm install is run with `NODE_ENV=production`, so only the `dependencies` will get installed. As such, anything required by `npm run build` will need to be in `dependencies` rather than `devDependencies`.

caseywilliams commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+coverage/
```
This is created when running the karma tests (same as under grunt) - I'll add a little note in this file saying so.

caseywilliams commented:
Fixing Bug 1317752 would absolutely help a lot with full stack development, since we could get rid of the watch task that's currently required to log into the unminified UI.

It wouldn't help with the Vagrant problems you mention, though. I guess the price of working with ES6 is that we can't just run the UI without doing some sort of node build beforehand. I don't think I have any idea of how to get around that for now, but I'll keep thinking.

New commit coming today with changes based on this feedback, thank you!

edmorley commented:
The Travis UI test chunk still takes a fair amount of time to run (5min10s). Would inverting the order of the `npm build` and `npm test` steps speed up the run? (Since the test step wouldn't need to rebuild?)

Not a big deal if not. Speeding things up is always something we can work on later (and might come with eg updating to newer node etc)

eliperelman commented:
@edmorley @caseywilliams so, I'll raise this now since this seems like the right time to do this.

Are you familiar with Yarn? It's an alternate npm client that is orders of magnitude faster than the npm client. It also brings deterministic installs, waaayyy better dependency pinning via always-in-sync yarn.lock file, and easily cacheable assets between Travis builds, so it goes much faster too. We use it exclusively for TC front-end, and want to push for its usage within all of R&P. It's almost a drop-in replacement.

I'd be happy to discuss more if you're interested in converting.

https://yarnpkg.com/en/

edmorley commented:
Yeah I've been following Yarn progress and am keen to switch to it soon (now that the worst bugs are resolved and both Travis and the Heroku nodejs buildpack have support). However for the Travis run speed here it won't actually make a huge difference, since we already use caching so the `npm install` is only 10s out of 5 mins.

edmorley commented:
Thank you for updating the PR :-)

Sorry for the churn by not being clearer before, but for splitting up the PR I meant to keep the related parts together (eg docs and the things they are documenting) but to move out linter/style fixes and anything else that could land prior to this work. 

eg something roughly like:

* **PR 1:** "Bug 1336556 - Preparation for switch to Neutrino/webpack"...
  - Commit A: "Bug 1336556 - Remove unused UI tests"
    - eg ` tests/ui/e2e/*`, `watchr.rb` and friends, removing the comment from `.travis.yml`.
  - Commit B: "Bug 1336556 - Make UI test naming more consistent"
    - With the current PR, the file adds and removals for this change are different commits, so git doesn't see them as renames but new files which breaks git log etc.
  - Commit C: "Bug 1336556 - Use explicit Angular dependency annotations"
    - ie the `ng-strict-di` and `treeherderApp.config(['$compileProvider', ...` additions - currently they make the diffs very noisy due to the indentation changes. 
  - Commit D: "Bug 1336556 - <something about JS style fixes>"
    - eg the `s/module/angular.mock.module/` changes, changing the `use strict` quotes style, the `_.unescape(...)` removal in reactrevisions.js (presuming it's unrelated to the rest of the changes?), whitespace changes.
  - Commit E: "Bug 1336556 - <something about HTML markup fixes>"
    - eg fixing tags that weren't closed properly (like in `thWatchedRepoNavPanel.html`, `comparechooserctrl.html`), adding new elements (such as `<span class="navbar-left">`) , whitespace changes including removing Windows line endings in `thActiveFiltersBar`.
  - Commit F: "Bug 1336556 - Switch to unminified version of jquery.flot.selection.js"
  - Commit G: "Bug 1336556 - Add Karma coverage directory to .gitignore"
  - ...plus anything else I've missed that is standalone cleanup.

* **PR 2:** (this one)
  - Commit A: "Bug 1336556 - Build ui with neutrino/webpack instead of grunt"
    - All other changes including docs and Gruntfile.js removal [since with the package.json changes Grunt won't work anyway], but not the ` ui/vendor/*` file removals.
  - Commit B: "Bug 1336556 - Remove unused vendored JS libraries"

edmorley commented on neutrino-custom/base.js:
```diff
@@ -0,0 +1,189 @@
+    neutrino.config.output.path(DIST);
+
+    if (process.env.NODE_ENV !== 'test') {
+        // Include files from node_modules in the separate, more-cacheable vendor chunk:
```
Rather than having to keep the list of vendor deps up to date, would something based on filepath like this work?
https://webpack.js.org/guides/code-splitting-libraries/#implicit-common-vendor-chunk

caseywilliams replied:
Hmm, yes -- that was my initial approach, but getting it to work with webpack-chain/Neutrino was nonobvious, although @eliperelman may have some ideas :). As I understand it, Neutrino sets up an empty commons chunk plugin and waits for you to add a vendor entrypoint like this that takes care of determining what will go into it. Using this mechanism, I don't think we can just do something like `reqiure('package.json').dependencies' to get a list of the right modules, since these `add(thing)` calls expect to find a javascript file eventually -- this means that for libs like bootstrap, which provide both css and js and have no default export, we're missing the right path for the js export by default (fontawesome was also a problem here because it's not even js).

I think another approach I could try is removing the default commons chunk plugin altogether and setting up another one based on the implicit scheme. I will give it a try today!

eliperelman replied:
Casey is right, this works only if you are using modules that work correctly in a CJS environment. This means packages like bootstrap and font-awesome will blow up your build if they aren't excluded. Using a whitelist approach avoids this at the cost of having some overhead maintenance.

edmorley replied:
Ah makes sense - thank you for the explanation :-)

caseywilliams commented:
Thanks for the detailed breakdown - I'll start splitting it up right now! Also, as you can see, something I've done (probably to the travis config) has broken the selenium tests yet again. They're still passing on my vagrant box, so any extra eyes on it would be very much appreciated, if anyone has time.

edmorley commented:
Some other things that occurred to me:

1. For users who run the Vagrant environment from a Windows host, trying to create symlinks within the guest OS in a directory that is mapped to the host (ie virtualbox shared folder) causes errors without some workarounds on the host side. In the past I seem to remember errors trying to npm install certain packages in the VM since they create symlinks in node_modules. 
  As such, we'll need to check if this occurs with the new package.json packages and figure out a workaround/documentation if so (at least two developers working on Treeherder use Windows environments for the host, one of which is myself).

2) neutrino-preset-web uses babel-preset-env to determine which browsers to target, and by default uses `last 2 {Chrome,Firefox,Edge,Opera,Safari,iOS} versions` ([here](https://github.com/mozilla-neutrino/neutrino-dev/blob/f5712521246c0bdf6e27d5f3abfd3f7625030e66/packages/neutrino-preset-web/src/index.js#L129-L135)). Given that (a) Treeherder doesn't work well on mobile anyway, (b) it's a developer targeted tool, so we shouldn't need to support older browsers, I was wondering whether we should override and use "last 1 X version" instead, plus drop iOS/Safari from the list?
  Judging by [this caniuse.com result](http://caniuse.com/#compare=edge+13,edge+14,firefox+50,firefox+51,chrome+55,chrome+56,safari+9.1,safari+10,opera+42,opera+43,ios_saf+9.3,ios_saf+10.0-10.2&compare_cats=JS%20API) it would presumably mean much fewer babel plugins have to be used, speeding up build time and reducing bundle size bloat (eg older versions of Safari/iOS Safari don't even support arrow functions or `let`). Might be worth measuring to see.

caseywilliams commented:
Ok, I've moved the react components back to their original locations and made the `last 1 X` version changes in the babel-preset-env config.

edmorley commented on neutrino-custom/base.js:
```diff
@@ -0,0 +1,204 @@
+        .test(/\.jsx?/)
+        .include(UI)
+        .loader('eslint', require.resolve('eslint-loader'), {
+            configFile: path.join(CWD, '.eslintrc')
```
Here and in a few other places a loader or plugin is replaced entirely, when I think the options could just be updated? (According to https://github.com/mozilla-rpweb/webpack-chain#configmodule)

@eliperelman is this correct?

edmorley replied:
Ah yes, like this:
https://github.com/mozilla-neutrino/neutrino-dev/blob/5c53dc04679d03aa6792449cab18639e75dcebdd/packages/neutrino-preset-react/src/index.js#L58-L70

edmorley replied:
Ah and looks like if we switch to the airbnb style preset (in the future, not in this PR), there's an even simpler way to do it:
https://neutrino.js.org/presets/neutrino-preset-airbnb-base/#eslintrc-config

(I think in the future I like the idea of overriding as few of the airbnb rules as possible, and just sticking with their style)

caseywilliams replied:
As far as I can tell, the base web/react presets don't ship with a linting rule by default (hence why they check for one with `rules.has('lint')`) -- so in this specific case, I think this is the first time a lint loader is created (but please correct me if I'm wrong!). Regardless, it seems like a better approach here would be to create a custom lint preset [as the docs describe](https://neutrino.js.org/presets/neutrino-lint-base/) and use that instead of this one-off. I'll do that today unless anyone has other suggestions.

I'll go through the custom presets in this PR and make sure all the loaders and plugins are modifying existing ones wherever possible, and get rid of any explicit requirements already covered by the neutrino presets.

eliperelman replied:
We started an official preset for rpweb projects, although it bakes in React preset. It uses Airbnb along with any rules we decide on during discussions. 

https://github.com/mozilla-rpweb/neutrino-preset-mozilla-rpweb

edmorley commented:
Testing locally and comparing the build output ([filelist](https://emorley.pastebin.mozilla.org/8980901)) against that produced by the existing grunt build ([filelist](https://emorley.pastebin.mozilla.org/8980900)) I see a few differences:
* The angular partials aren't built
* These files are missing: robots.txt, revision.txt, contribute.json
* The fonts are included twice, once hashed in dist/ and once in dist/vendor/fonts/

I've been working on bug 1343624, though the symlink situation is adding complication (eg https://emorley.pastebin.mozilla.org/8980904) - though I will be able to come up with a solution, it will just be hacky (annoyingly I can see a change or two that npm could do to make this easier, but upstreaming something is another rabbit hole that will have to be saved for later...).

caseywilliams commented:
The missing *.txt and *.json files and the doubled-up fonts were silly mistakes in the CopyPlugin config, but bundling the angular partials in a way that works for both directive `templateUrl: ...` and html `ng-include` situations seems a little more difficult. There are a few webpack loaders that are meant to just load templates into the template cache at build time -- this [wiki article](https://github.com/dmachat/angular-webpack-cookbook/wiki/Angular-Template-Cache) seems promising. I'll try setting it up. Again, thanks so much for your help!

caseywilliams commented:
I've added a third commit which loads the html partials into the Angular template cache when the various `ng-app`s bootstrap. This seems to work fine, but there's a shortcoming in that the contents of the templates are loaded and cached for each one of the `ng-app`s. It would be nice to have this happen only once, but I couldn't figure out how to 'share' a run block between multiple apps -- I'd love to hear any other suggestions on how to go about this!

edmorley commented:
Thank you for the update :-)

Building again I see `robots.txt` is now present, but `contribute.json` (which was copied in from the repo root rather than `ui/`) is still missing. There's also a `partials/perf/bugzilla_talos.tmpl` in the generated output, and duplicate svg font files under `vendor/fonts/`.

When building I received `exit status 137` errors unless I stopped mysql/elasticsearch - looks like it was running out of RAM (using nodejs 7.7.1 at least)? I've filed bug 1344983 to increase the Vagrant instance to something more than 2GB.

Also when building, the command didn't exit 0 due to eslint errors:
https://emorley.pastebin.mozilla.org/8981243
...fixing these is likely out of scope for this bug, but it would be good to mark these as allowed to fail for now, so at least `npm run build` doesn't display errors.

edmorley commented on neutrino-custom/base.js:
```diff
@@ -0,0 +1,205 @@
+        // Ensure chunkhashes for unchanged chunks don't change -- this allows the bundled vendor
+        // file to keep its hash and stay cached in the user's browser when it hasn't been updated.
+        neutrino.config.output.chunkFilename('[chunkhash].[id].chunk.js');
+        neutrino.config.plugin('hash').use(Md5HashPlugin);
```
Reading https://webpack.js.org/guides/caching/#generating-unique-hashes-for-each-file left me slightly confused. The docs first give the impression hashed filenames are supported out of the box, then go on to suggest that the default way of generating the hashes causes unnecessary invalidation in some cases, without being very clear as to what they are.

Either way, this seems like something that Neutrino could/should be handling for us. @eliperelman do you agree? If so, would you mind filing a ticket against it?

eliperelman replied:
I'll open it up for discussion on Neutrino and see if there is any consensus.

caseywilliams commented:
Okay, I think I may have finally addressed all of these - thank you for your patience, @edmorley!

The eslint issue perplexes me - I'm not seeing any linting failures on running `npm run build` myself, but if it's okay to leave it until later, then I guess the point is moot for now.

edmorley commented on package.json:
```diff
@@ -10,40 +10,52 @@
   },
   "scripts": {
-    "test": "karma start tests/ui/config/karma.conf.js --single-run --browsers Firefox"
+    "build": "neutrino build --presets ./neutrino-custom/production.js",
```
We'll need to use the relative paths here due to the use of `--no-bin-links` by #2243.
In the case of neutrino iirc that will be `./node_modules/neutrino/bin/neutrino`.

eliperelman replied:
Correct, although its sad that this has to be done. :cry:

edmorley replied:
Agreed -- though when I get a moment I'm going to try proposing (in both the npm and yarn projects) that `--no-bin-links` is changed so it does what happens on Windows (which is creating shell scripts that call the real script) rather than the current behaviour of "not create anything in the .bin directory at all". We'll just have to see how well that idea goes down...

eliperelman commented on ui/js/cache-templates.js:
```diff
@@ -0,0 +1,23 @@
+// This is a run block which, when used on an angular module, loads all
+// of the partials in /ui/partials and /ui/plugins into the Angular
+// template cache. This means that ng-includes and templateUrls will not
+// actually need to request a partial file at any point.
```
Unfortunately I know nothing about Angular nor these loaders :/

caseywilliams replied:
I attempted to use ngtemplate-loader and ng-cache-loader before trying this approach - they worked in cases where the template was included via `templateUrl: require('....html')`, but I couldn't make them work with `ng-include` in the main templates (index.html etc.) which are rendered by the html plugin. It may be possible, but I just couldn't find a way to make them work together successfully.

caseywilliams commented:
I've reworked these commits to use yarn instead of npm and moved the eslint configuration  away from the `.eslintrc` file and into the neutrino linting preset in `neutrino-custom/lint.js`, but I still have one major problem - eslint seems to behave differently when run directly via its CLI vs through yarn. For example, if I run `eslint ui/` to use the configuration generated by `.eslintrc.js`, there are no errors, but if I run any of the yarn tasks that incorporate eslint, they report indentation errors that aren't present using the former method. 

I was able to reproduce these errors using the eslint CLI by changing the [`indent` rule](http://eslint.org/docs/rules/indent) to add a `parameters: 'first'` requirement. So it seems like even though the default is to _not_ care about parameter indentation, this rule may adjusted somehow somewhere in the configuration. I'm not really sure where to begin solving this.

I've also updated the documentation, but after I've fixed this linting problem I'll need go back and run through all of the instructions in the docs to make sure they're still accurate.

eliperelman commented:
AHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH!

I tried fixing this all day and couldn't figure it out.

Then I ran:

```bash
yarn upgrade eslint && ./node_modules/.bin/eslint ui/**/*.js
```

AND NOW IT WORKS.

![rage](http://gif.co/x8u9.gif)

So, looks like the eslint in package.json is an older version, which isn't the same as the eslint version used by eslint-loader and Neutrino. And I'm guessing the older one is b0rked. I'd recommend either upgrading the eslint in package.json, or removing it altogether and using the version that comes along with eslint-loader.

Lemme know if that works.

eliperelman commented on neutrino-custom/lint.js:
```diff
@@ -0,0 +1,90 @@
+                },
+                extends: 'eslint:recommended',
+                rules: {
+                    'accessor-pairs': 2,
```
Something that would help the readability of these rules would be to replace all `2`s with `'error'`.

caseywilliams commented:
Okay, whew - I think this is ready for another look, @edmorley - I updated eslint and fixed the new errors in a separate commit before the webpack conversion. Thanks so much to you and Eli for all your help in working through these changes!

edmorley commented:
Thank you for the latest changes! I'm on vacation at the moment, back Tuesday.

In the meantime, @camd, @wlach, @jgraham would you mind checking this out and testing how you find the various development workflows (eg full stack, UI-only against prod API, running JS tests etc) and whether you feel the docs need any tweaks? 

Note: If your Vagrant instance hasn't recently been recreated, you'll need to do a `vagrant reload` to pick up the increase to 3GB RAM. In addition, if you have a `node_modules` directory sitting around from a while ago, you might need to purge that.

camd commented:
Hey all—  I’m running out of time today and I’m off Friday as well.  But I’ll check this out on Monday when I’m back.  I tried this a bit and so far so good.  I definitely had to clear out some cruft as I was getting errors earlier.  But the command Eli gave in his “angry Leo DiCaprio” comment helped..  :)

I’ll play more with it Monday.

Thanks Casey!
-Cam


> On Mar 22, 2017, at 9:22 AM, Ed Morley <notifications@github.com> wrote:
> 
> Thank you for the latest changes! I'm on vacation at the moment, back Tuesday.
> 
> In the meantime, @camd <https://github.com/camd>, @wlach <https://github.com/wlach>, @jgraham <https://github.com/jgraham> would you mind checking this out and testing how you find the various development workflows (eg full stack, UI-only against prod API, running JS tests etc) and whether you feel the docs need any tweaks?
> 
> Note: If your Vagrant instance hasn't recently been recreated, you'll need to do a vagrant reload to pick up the increase to 3GB RAM. In addition, if you have a node_modules directory sitting around from a while ago, you might need to purge that.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/mozilla/treeherder/pull/2159#issuecomment-288454840>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAZoVFHmj_7pV1aYd0JCzXUs88LXkTgRks5roUrFgaJpZM4L_XBr>.
>

camd commented:
This works great for me.  I *did* need to nuke my node_modules folder to get rid of cruft.  But after that, I tried the various ``yarn run start:<local>|<stage|etc`` flavors and things did what I expected.  The ``yarn run test:watch`` also worked well, as expected.  In addition, by the time I'd finished testing, my hiccups were gone.  So... BONUS!

This seems good to land by my reckoning.  :)

edmorley commented:
After nuking my existing node_modules directory, I get this warning when yarn installing:
`warning "eslint-plugin-react@6.10.0" has unmet peer dependency "eslint@^2.0.0 || ^3.0.0".`

There's also now no `yarn run lint` command - I know some of the other commands do linting, but perhaps we need an easy way to run eslint on it's own?

However those aside (I think they can wait until followups) I think we're good to go here once rebased + conflict fixed in `ui/js/controllers/main.js` :-)

Thank you for persevering through this!

caseywilliams commented:
Ok, I've rebased and fixed a small linting error in `main.js`. I also went ahead added a new commit that and installs eslint explicitly (instead of relying on the neutrino dependency), which should get rid of the install warnings, and added a new eslint task that just does linting (`yarn run lint`).
